### PR TITLE
Timing and performance archiving modifications by Pat Worley

### PR DESCRIPTION
**Pull request by jeff-cohere**
These are the modifications that Pat Worley and I have been working on in our fork of CIME. This is ongoing work that will be continued in Rob Jacob's ACME CIME merge branch.

### CODE REVIEW

**Comment by jeff-cohere** [comment]
@rljacob, give me a sec to rebase this so I can fix the author entries on some of these commits.

**Comment by jeff-cohere** [comment]
ok, should be ready to go. Do you want me to perform this merge?

**Comment by rljacob** [comment]
Yes go ahead.

**Comment by rljacob** [comment]
Actually, let me look at this a bit.

**Comment by rljacob** [comment]
Does every machine have to create those postrun/prerun scripts?

**Comment by jeff-cohere** [comment]
It's not clear to me exactly how that mechanism works. The logic that executes this stuff is in machines-acme/template.cesmrun, in runCESM(), lines 312 and 329. It seems like @worleyph needs them for those machines he is using to collect performance numbers, but it's not clear that if they aren't present that they will halt the run. Pat, if you're monitoring your email, can you comment?

**Comment by jeff-cohere** [comment]
You might try running the coupler tests on blues in this branch to see whether the pre/postrun stuff trips it up, Rob.

**Comment by worleyph** [comment]
> Does every machine have to create those postrun/prerun scripts?
> ...
> but it's not clear that if they aren't present that they will halt the run.

create_newcase looks for a system-specific prerun/postrun script. If one exists, it will use this. Otherwise if will use the "generic" script. At the moment, there is only full support for Edison. Once this is working on Titan and Mira, I'll add these back in, as well as support for any other system that we are interested in.

This mimics the logic in ACME currently, and is simply meant to support the current functionality so that we can base further modifications on a version that does the right thing.

**Comment by rljacob** [comment]
What do I look for to see if the generic script is working?  I'm trying this on blues.

**Comment by worleyph** [comment]
I believe that if it configures, builds and runs, it works. The generic scripts are meant to do nothing, or, at least, do no harm. There is a little handshaking between the scripts that requires some minimal logic even for generic. Note that even this handshaking might be removable with a different implementation of the logic.

**Comment by rljacob** [comment]
Ok.  The drv tests all pass on Blues so I'll go ahead and merge this.

File: scripts/create_newcase [hunk 1]
```diff
@@ -1060,6 +1066,53 @@ if( -e "$machdir/Depends.$compiler" ) {
...
 }
+}
+
+# The following creates cesm_postbuild, cesm_prerun, cesm_postrun, cesm_postrun2, and mach_syslog
+# The following creates cesm_postbuild
+if( -e "$scriptsroot/Tools/cesm_postbuild" ) {
+  $sysmod = "cp $scriptsroot/Tools/cesm_postbuild $caseroot/Tools/cesm_postbuild";
+  system($sysmod) == 0 or die "ERROR: $sysmod failed: $?\n";
+}else{
+  $sysmod = "cp $scriptsroot/Tools/cesm_postbuild.noop $caseroot/Tools/cesm_postbuildn";
```
Comment (rljacob) [inline scripts/create_newcase]: Is that a typo at the end of line 1077?   "cesm_postbuildn" ?

**Comment by rljacob** [comment]
It looks like this checkin also adds the PIO runtime options.  That's fine but want to note it for the commit message.   Also what all are the  postbuild,prerun,postrun scripts supposed to do?

File: scripts/create_newcase (hunk 1 above)
Comment (worleyph) [inline scripts/create_newcase]: Yes. This is a typo, and it came from me (is in my version also). This is an if-test branch that is not currently being executed, and that I added as an example of what we might want to do if releasing this without the performance archiving logic.

Should be

  $sysmod = "cp $scriptsroot/Tools/cesm_postbuild.noop $caseroot/Tools/cesm_postbuild";

**Comment by worleyph** [comment]
>  That's fine but want to note it for the commit message. Also what all are the postbuild,prerun,postrun scripts supposed to do?

Note that I did not make these commits nor this pull request. I will be happy to work with Jeff to expain the parts that I am the author of, but will wait for Jeff to ask what he wants from me, and in what form. I did mention the PIO support to Jeff, some of which is commented out so that it does not depend on these PIO options being in ACME CIME, though the expectation is that they will be at some point?

**Comment by rljacob** [comment]
Draft commit message.  Feel free to edit this comment.

Add more ACME features to CIME
- Add ACME timers to driver.
- Add first-pass of ability to archive enhanced timing and provenance information.
- modify cesm_build.pl to save build times and call:
  - cesm_postbuild - saves git and svn info, sourcemods and build_environment to $exe_dir.
- modify template.cesmrun to call:
  - cesm_prerun.$mach - after job is queued, add queue info to CaseStatus, archive $caseroot info, start syslog
  - cesm_postrun.$mach - after job runs, save jobid info
  - cesm_postrun2.$mach - archive logs, and info from postbuild.
  - archive location is $save_timing_dir/performance_archive/$user/$case/$lid
- Add ability to use Depends.mach.compiler
- Add ACME's pio runtime options (commented out for now).
- Add syslog.mach progress monitoring scripts.
- Update GPTL with latest ACME version.

**Comment by rljacob** [comment]
I'm integrating this PR, hence the questions.   Yes we will be using the PIO that uses those options.

**Comment by jeff-cohere** [comment]
Thanks for elaborating on the commit message. I don't think I can add more. Typo is fixed. @rljacob, is there anything else you think we need to do to proceed?

**Comment by worleyph** [comment]
@rljacob, Thanks! I was unable able to get to this task last night, and your commit messages look very complete. I'll glance at the code mods by 1pm Eastern today and see if there is anything else worth mentioning.

**Comment by rljacob** [comment]
Following Steve's advice, I'm going to update master with some changes from the subtree in my branch, then merge this branch, then update the subtree.

**Comment by worleyph** [comment]
A little more detailed commit messages, if interested. @rljacob suggestions are fine by me:

Add more ACME features to CIME
- Add ACME timers and timing logic to driver, including altered treatment of initialization, saving per process data for each component root process, and adding support for an option to collect only the minimum-required performance data.
- Add ACME enhanced control of performance data collection:
  - prof_inparm namelist variables: profile_outpe_num, profile_outpe_stride, profile_ovhd_measurement
  - papi_inparm namelist variables: papi_ctr1_str, papi_ctr2_str, papi_ctr3_str, papi_ctr4_str
  - env_run.xml variables: TIMER_DETAIL (for profile_timer_detail), TPROF_TOTAL (number of performance data checkpoints)
- Add first-pass of ability to archive enhanced timing and provenance information, with full support for Edison only.
- modify cesm_build.pl to save build times and call:
  - cesm_postbuild - saves git and svn info, sourcemods and build_environment to $exe_dir.
- modify template.cesmrun to call:
  - cesm_prerun.$mach - after job is queued, add queue info to CaseStatus, archive $caseroot info, start syslog
  - cesm_postrun.$mach - after job runs, save jobid info, kill syslog job (if necessary)
  - cesm_postrun2.$mach - archive logs, and info from postbuild.
  - archive location is $save_timing_dir/performance_archive/$user/$case/$lid
- Add ability to use Depends.mach.compiler
- Add ACME's pio runtime options (commented out for now).
  - env_run.xml variables (commented out): PIO_REARR_COMM_TYPE, PIO_REARR_COMM_FCD, PIO_REARR_COMM_MAX_PEND_REQ, PIO_REARR_COMM_ENABLE_HS, PIO_REARR_COMM_ENABLE_ISEND
  - pio_default_inparm namelist variables: pio_rearr_comm_enable_hs, pio_rearr_comm_enable_isend, pio_rearr_comm_fcd, pio_rearr_comm_max_pend_req, pio_rearr_comm_type
- Add syslog.mach progress monitoring scripts.
- Update GPTL with latest ACME version.
  - and also modify perf_mod.F90 to adopt element of API from CIME version so as to be consistent with usage in driver routines. This disables some functionality, but does not disable required functionality. A future check-in will restore this in a way consistent with both CIME and ACME.

**Comment by worleyph** [comment]
@rljacob , since this merge is being delayed a little, should we add support for performance archiving on Mira and Edison before the merge? Also, these changes are NOT compatible with current ACME POP and CICE, but only a simple modification is required in POP and CICE to support building and running. Timing overhead will increase until the ACME version of GPTL is modified to fully support the new API.

**Comment by rljacob** [comment]
You mean Mira and Titan?  I'm going to merge this by tomorrow.

**Comment by worleyph** [comment]
If you are going merge tomorrow (before it is working on Titan), then no reason to wait. I do want Mira and Titan support in there before this becomes operational for ACME though. Thanks.

**Comment by rljacob** [comment]
I meant I'm merging this branch in to cime and then in to my ACME branch.  We're still a week away from merge to ACME master.

**Comment by worleyph** [comment]
Understood. I just want to get Titan and Mira support into your branch before the  merge to ACME master. Please tell me what works best for you.

**Comment by rljacob** [comment]
After merging this to master locally on blues, all drv tests pass.

**Pull request merged**